ui-core/src/components/inputs/TolerancePicker.tsx

```diff
@@ -0,0 +1,145 @@
+  onToleranceChange: (toleranceValues: ToleranceValues) => void;
+  toleranceValues?: ToleranceValues;
+  translateWarningMessage?: (message: string) => string;
+};
+
+const TolerancePicker: React.FC<TolerancePickerProps> = ({
+  onToleranceChange,
+  toleranceValues: { minusTolerance, plusTolerance } = {
+    minusTolerance: TOLERANCE_RANGES[0].value,
+    plusTolerance: TOLERANCE_RANGES[0].value,
+  },
+  translateWarningMessage,
+  ...inputProps
+}) => {
+  const formatToleranceValue = (minusIndex: number, plusIndex: number) =>
+    `-${TOLERANCE_RANGES[minusIndex]?.label || ''}/+${TOLERANCE_RANGES[plusIndex]?.label || ''}`;
+
+  const [showPicker, setShowPicker] = useState(false);
+  const [inputValue, setInputValue] = useState(formatToleranceValue(0, 0));
+  const [warningStatus, setWarningStatus] = useState<StatusWithMessage | undefined>(undefined);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    const minusToleranceIndex = TOLERANCE_RANGES.findIndex(
+      (range) => range.value === minusTolerance
+    );
+    const plusToleranceIndex = TOLERANCE_RANGES.findIndex((range) => range.value === plusTolerance);
+    if (minusToleranceIndex < 0 || plusToleranceIndex < 0) {
+      const invalidTolerance = minusToleranceIndex < 0 ? minusTolerance : plusTolerance;
+      setWarningStatus({
+        status: 'warning',
+        message:
+          translateWarningMessage?.(invalidTolerance.toString()) ||
+          `${invalidTolerance} is not a valid tolerance value.`,
+      });
+    } else {
+      setWarningStatus(undefined);
+    }
+
+    setInputValue(formatToleranceValue(minusToleranceIndex, plusToleranceIndex));
+  }, [minusTolerance, plusTolerance]);
+
+  return (
+    <div className="tolerance-picker">
+      <div>
+        <Input
+          {...inputProps}
+          value={inputValue}
+          statusWithMessage={warningStatus}
+          readOnly
+          onClick={() => setShowPicker(!showPicker)}
+          type="text"
+          ref={inputRef}
+        />
+      </div>
+
+      <Modal inputRef={inputRef} isOpen={showPicker} onClose={() => setShowPicker(false)}>
+        <div className="time-tolerance">
+          <div className="time-picker-section">
+            <div className="tolerance-grid">
+              {TOLERANCE_RANGES.map(({ label, value }, index) => {
+                return (
+                  <button
+                    key={`${label}-${index}`}
+                    className={cx('minute', { selected: minusTolerance === value })}
+                    style={{ gridArea: `a${index + 1}` }}
+                    onClick={() =>
+                      onToleranceChange({
+                        minusTolerance: value,
+                        plusTolerance,
+                      })
+                    }
+                  >
+                    {`-${label}`}
+                  </button>
+                );
+              })}
+            </div>
+          </div>
```

Comment:
Here’s a suggestion: since these lines are used to render a range button, we can extract this into a component. This way, we can reuse it for both minus and plus ranges. What do you think?